### PR TITLE
feat(ctb): `ctb-test-case-generator` package

### PR DESCRIPTION
**Pull request by clabby**
# Overview

Adds a fuzz test input generator for testing Bedrock's `MerkleTrie.sol`.

Go version of #3847

**Metadata**
ENG-3048

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* sc/ctb-mini-trie-refactor
  * **PR #4233** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4233" target="_blank">Graphite</a>  👈
    * **PR #4234** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4234" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/4233?utm_source=stack-comment).

**Review by maurelian — APPROVED** [review]
LGTM

File: packages/contracts-bedrock/go-fuzz/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+A lightweight input fuzzing utility used for testing various Bedrock contracts.
+
+<pre>
+├── go-fuzz
+│   ├── <a href="./cmd">cmd</a>: `go-fuzz`'s binary
+│   └── <a href="./trie">trie</a>: Utility for generating random merkle trie roots / inclusion proofs
+</pre>
+
+## Usage
+
```
[the remark this answers is not in the document]
Comment (trianglesphere) [reply packages/contracts-bedrock/go-fuzz/README.md:12]: The build fuzz command is not included in this PR

[the remark this answers is not in the document]
Comment (trianglesphere) [reply packages/contracts-bedrock/go-fuzz/README.md:12]: Either this or the next PR should have this addition to the docs.

[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/go-fuzz/README.md:12]: Will add this in the contracts changes PR where the `build:fuzz` script was added after @maurelian's changes are up- he said he's planning on renaming the tests and pulling out the formatting changes this morning.

File: packages/contracts-bedrock/go-fuzz/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# `go-fuzz`
```
Comment (trianglesphere) [inline packages/contracts-bedrock/go-fuzz/README.md:1]: I think this package / binary should be renamed.

**Review by trianglesphere — COMMENTED** [review]
I'm bumping a comment of mine from the last version of this PR. I don't particularly like the name `go-fuzz` because it conflicts with several existing go based projects & doesn't capture what this program does - generate trie (or other types of) test cases.

File: packages/contracts-bedrock/go-fuzz/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/go-fuzz/README.md:12]: #4319 was just merged

File: packages/contracts-bedrock/go-fuzz/README.md (hunk 2 above)
Comment (clabby) [reply packages/contracts-bedrock/go-fuzz/README.md:1]: Good point- sorry that this feedback got lost, there was some confusion on whether or not @smartcontracts' trie changes were going to be merged and this PR has been closed/reopened a few times.

Any suggestions? Maybe one of:
* `go-ctb-fuzz`
* `go-ctb-fuzzer`
* `ctb-fuzz`

Comment (trianglesphere) [reply packages/contracts-bedrock/go-fuzz/README.md:1]: I lean to something along the lines of `ctb-test-case-generator`. There's no reason to include the word "go" in this. `ctb-fuzz` is okayish, but I don't actually think of this code as a fuzzer.

File: packages/contracts-bedrock/go-fuzz/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/go-fuzz/README.md]: Added in #4234

File: packages/contracts-bedrock/go-fuzz/README.md (hunk 2 above)
Comment (clabby) [reply packages/contracts-bedrock/go-fuzz/README.md]: Name changed to `ctb-test-case-generator`

**Comment by clabby** [comment]
@mergify ignore

**Comment by trianglesphere** [comment]
> > I'm bumping a comment of mine from the last version of this PR. I don't particularly like the name `go-fuzz` because it conflicts with several existing go based projects & doesn't capture what this program does - generate trie (or other types of) test cases.
> 
> What do you think of the name `op-contracts-fuzzer`?
> 
> edit: i see you have changed the name already, you don't need to change it again if you don't want to

I agree with Mark's edit, but `op-contracts-fuzzer` is a pretty good name. I like how it fits in with the rest of the `op-*` go stuff.

**Comment by clabby** [comment]
@tynes @trianglesphere Can change the name again quickly, so no problem there- Also agreed w/ @trianglesphere's earlier comment about `*-fuzz` / `*-fuzzer` being a misleading name given that the package solely generates test cases. Maybe `op-contracts-tcg` (`tcg` = Test Case Generator`)?

**Review by tynes — APPROVED** [review]

**Comment by tynes** [comment]
I don't think this needs a `mergify-ignore` tag anymore, it should be good to merge

**Pull request merged**